framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,839 @@
+                // This is currently not supported. At some point we might add
+                // polymorphic qualifiers with upper and lower bounds.
+                throw new UserError(
+                        "AnnotatedTypeFactory: "
+                                + qualifierKind
+                                + " is polymorphic and specifies super qualifiers.%n"
+                                + "Remove the @PolymorphicQualifier or @SubtypeOf annotation from it.");
+            } else if (!isPoly && !hasSubtypeOfAnno) {
+                throw new UserError(
+                        "AnnotatedTypeFactory: %s does not specify its super qualifiers.%n"
+                                + "Add an @SubtypeOf or @PolymorphicQualifier annotation to it,%n"
+                                + "or if it is an alias, exclude it from `createSupportedTypeQualifiers()`.%n",
+                        qualifierKind);
+            } else if (isPoly) {
+                if (qualifierKind.top == null) {
+                    throw new UserError(
+                            "PolymorphicQualifier, %s,  has to specify type hierarchy, if more than one exist; top types: [%s] ",
+                            qualifierKind, SystemUtil.join(", ", tops));
+                } else if (!tops.contains(qualifierKind.top)) {
+                    throw new UserError(
+                            "Polymorphic qualifier %s has invalid top %s. Top qualifiers: %s",
+                            qualifierKind, qualifierKind.top, SystemUtil.join(", ", tops));
+                }
+            }
+        }
+
+        if (bottoms.size() != tops.size()) {
+            throw new UserError(
+                    "Number of tops not equal to number of bottoms: Tops: [%s] Bottoms: [%s]",
+                    SystemUtil.join(", ", tops), SystemUtil.join(", ", bottoms));
+        }
+    }
+
+    /**
+     * Creates all QualifierKind objects for the given qualifier classes and adds them to
+     * qualifierClassMap. This method does not initialize all fields in the {@link QualifierKind};
+     * that is done by {@link #initializeQualifierKindFields(Map)}.
+     *
+     * @param qualifierClasses classes of annotations that are type qualifiers
+     * @return a mapping from the canonical name of an annotation class to {@link QualifierKind}
+     */
+    protected NavigableMap<@Interned String, QualifierKind> createQualifierKinds(
+            Collection<Class<? extends Annotation>> qualifierClasses) {
+        TreeMap<@Interned String, QualifierKind> nameToQualifierKind = new TreeMap<>();
+        for (Class<? extends Annotation> clazz : qualifierClasses) {
+            @SuppressWarnings("interning") // uniqueness is tested immediately below
+            @Interned QualifierKind qualifierKind = new QualifierKind(clazz);
+            if (nameToQualifierKind.containsKey(qualifierKind.name)) {
+                throw new UserError("Duplicate QualifierKind " + qualifierKind.name);
```

Comment:
You suggested adding a TypeSystemError to complement BugInCF and UserError.  I think that is a good thing to do, in a separate pull request that is merged before this one.